File: cpp/RingData.cpp
#include "RingData.h"

using namespace std;


RingData RingData::load(const string& filename,
                        bool metadata_only) {
    gz_file = GzipFile(filename, "rb");
    // Python 2.6 GzipFile doesn't support BufferedIO
    if (hasattr(gz_file, "_checkReadable")) {
        gz_file = BufferedReader(gz_file);
    }

    // See if the file is in the new format
    magic = gz_file.read(4);
    if (magic == "R1NG") {
        int format_version, = struct.unpack("!H", gz_file.read(2));
        if (format_version == 1) {
            ring_data = cls.deserialize_v1(
                gz_file, metadata_only=metadata_only);
        } else {
            raise Exception("Unknown ring format version %d" %
                            format_version);
        }
    } else {
        // Assume old-style pickled ring
        gz_file.seek(0);
        ring_data = pickle.load(gz_file);
    }

    if (not hasattr(ring_data, "devs")) {
        ring_data = RingData(ring_data["replica2part2dev_id"],
                             ring_data["devs"], ring_data["part_shift"]);
    }
    return ring_data;
}
 
RingData::RingData(RingStructure ring) {
}

/*
void RingData::serialize_v1(FILE* file_obj) {
    // Write out new-style serialization magic and version
    file_obj.write(struct.pack("!4sH", "R1NG", 1));
    ring = this->to_dict();
    json_encoder = json.JSONEncoder(sort_keys=true);
    json_text = json_encoder.encode(
        {"devs": ring["devs"], "part_shift": ring["part_shift"],
         "replica_count": len(ring["replica2part2dev_id"])});
    json_len = len(json_text);
    file_obj.write(struct.pack("!I", json_len));
    file_obj.write(json_text);
    for (part2dev_id in ring["replica2part2dev_id"]) {
        file_obj.write(part2dev_id.tostring());
    }
}
*/

/*
void RingData::save(const string& filename) {
    // Override the timestamp so that the same ring data creates
    // the same bytes on disk. This makes a checksum comparison a
    // good way to see if two rings are identical.
    tempf = NamedTemporaryFile(dir=".", prefix=filename, delete=False);
    gz_file = GzipFile(filename, mode='wb', fileobj=tempf, mtime=mtime);
    self.serialize_v1(gz_file);
    gz_file.close();
    tempf.flush();
    os.fsync(tempf.fileno());
    tempf.close();
    os.chmod(tempf.name, 0o644);
    os.rename(tempf.name, filename);
}
*/

File: scala/RingData.scala


/**
 * Partitioned consistent hashing ring data (used for serialization).
 */
class RingData {

    _ring: RingStructure
    devs: Array[StorageDevice]
    _replica2part2dev_id: Array[Array[Int]]
    _part_shift: Int = 0


    /**
     *
     * @param replica2part2dev_id
     * @param devs
     * @param part_shift
     */
    public RingData(RingStructure ring) {
        this._ring = ring;
    }

    /**
     *
     * @param gz_file
     * @return
     */
    public static RingStructure deserialize_v1(GZIPInputStream gz_file) {
        return RingData.deserialize_v1(gz_file, false);
    }

    /**
     * Deserialize a v1 ring file into a dictionary with `devs`, `part_shift`,
     *  and `replica2part2dev_id` keys.
     *
     * If the optional kwarg `metadata_only` is True, then the
     *  `replica2part2dev_id` is not loaded and that key in the returned
     *  dictionary just has the value `[]`.
     *
     * @param file gz_file An opened file-like object which has already
     *                       consumed the 6 bytes of magic and version.
     * @param bool metadata_only: If True, only load `devs` and `part_shift`
     * @return A dict containing `devs`, `part_shift`, and
     *            `replica2part2dev_id`
     */
    public static RingStructure deserialize_v1(GZIPInputStream gz_file, boolean metadata_only) {

        return RingStructure.deserialize_v1(filename);
    }

    /**
     *
     * @param filename
     * @return
     */
    public static RingData load(String filename) {
        return RingData.load(filename, false);
    }

    /**
     * Load ring data from a file.
     *
     * @param filename Path to a file serialized by the save() method.
     * @param metadata_only If true, only load `devs` and `part_shift`.
     * @return A RingData instance containing the loaded data.
     */
    public static RingData load(String filename, boolean metadata_only) {
        return RingStructure.deserialize_v1(filename);
    }

    /*
    public void serialize_v1(file_obj) {
        this.ring.serialize_v1(filename);
    }
    */

    /*
    public void save(String filename) {
        save(filename, 1300507380);
    }
    */

    /**
     * Serialize this RingData instance to disk.
     *
     * @param filename File into which this instance should be serialized.
     * @param mtime time used to override mtime for gzip, default or None
     *                 if the caller wants to include time
     */
    /*
    public void save(String filename, mtime=1300507380.0) {
        // Override the timestamp so that the same ring data creates
        // the same bytes on disk. This makes a checksum comparison a
        // good way to see if two rings are identical.
        tempf = NamedTemporaryFile(dir=".", prefix=filename, delete=false);
        gz_file = GzipFile(filename, mode="wb", fileobj=tempf, mtime=mtime);
        this.serialize_v1(gz_file);
        gz_file.close();
        tempf.flush();
        os.fsync(tempf.fileno());
        tempf.close();
        os.chmod(tempf.name, 0o644);
        os.rename(tempf.name, filename);
    }
    */

    /**
     *
     * @return
     */
    /*
    public Map<String,String> to_dict() {
        return {"devs": this.devs,
                "replica2part2dev_id": this._replica2part2dev_id,
                "part_shift": this._part_shift}
    }
    */
}

